src/execution/operator/order/physical_order.cpp

```diff
@@ -1,105 +1,882 @@
+	vector<unique_ptr<RowChunk>> var_sorting_sizes;
+
+	//! Payload data (and payload entry sizes if there is variable size data)
+	unique_ptr<RowChunk> payload_block = nullptr;
+	unique_ptr<RowChunk> sizes_block = nullptr;
+
+	//! Constant buffers allocated for vector serialization
+	const SelectionVector *sel_ptr = &FlatVector::INCREMENTAL_SELECTION_VECTOR;
+	data_ptr_t key_locations[STANDARD_VECTOR_SIZE];
+	data_ptr_t validitymask_locations[STANDARD_VECTOR_SIZE];
+	idx_t entry_sizes[STANDARD_VECTOR_SIZE];
+};
+
+template <class T>
+static idx_t TemplatedGetSize(Value min, Value max) {
+	T min_val = min.GetValue<T>();
+	T max_val = max.GetValue<T>();
+	idx_t size = sizeof(T);
+	T max_in_size = (1 << ((size - 1) * 8 - 1)) - 1;
+	while (max_val < max_in_size && min_val > -max_in_size) {
+		size--;
+		max_in_size = (1 << ((size - 1) * 8 - 1)) - 1;
+	}
+	return size;
+}
+
 unique_ptr<GlobalOperatorState> PhysicalOrder::GetGlobalState(ClientContext &context) {
-	return make_unique<OrderByGlobalOperatorState>();
+	auto &buffer_manager = BufferManager::GetBufferManager(context);
+	auto state = make_unique<OrderGlobalState>(buffer_manager);
+
+	// init sorting state and sorting block
+	size_t entry_size = 0;
+	vector<OrderType> order_types;
+	vector<OrderByNullType> order_by_null_types;
+	vector<LogicalType> types;
+	vector<BaseStatistics *> stats;
+	vector<bool> has_null;
+	vector<bool> constant_size;
+	vector<idx_t> col_sizes;
+	for (auto &order : orders) {
+		// global state ExpressionExecutor
+		auto &expr = *order.expression;
+		state->executor.AddExpression(expr);
+
+		// sorting state
+		order_types.push_back(order.type);
+		order_by_null_types.push_back(order.null_order);
+		types.push_back(expr.return_type);
+		if (expr.stats) {
+			stats.push_back(expr.stats.get());
+		} else {
+			stats.push_back(nullptr);
+		}
+
+		// compute column sizes
+		auto physical_type = expr.return_type.InternalType();
+		constant_size.push_back(TypeIsConstantSize(physical_type));
+		idx_t col_size = GetTypeIdSize(expr.return_type.InternalType());
+
+		// TODO: make use of statistics
+		if (!TypeIsConstantSize(physical_type)) {
+			switch (physical_type) {
+			case PhysicalType::VARCHAR:
+				col_size = StringStatistics::MAX_STRING_MINMAX_SIZE;
+			default:
+				// do nothing
+				break;
+			}
+		}
+		has_null.push_back(true);
+
+		// increment entry size with the column size
+		if (has_null.back()) {
+			col_size++;
+		}
+		entry_size += col_size;
+		col_sizes.push_back(col_size);
+
+		// create RowChunks for variable size sorting columns in order to resolve
+		if (TypeIsConstantSize(physical_type)) {
+			state->var_sorting_blocks.push_back(nullptr);
+			state->var_sorting_sizes.push_back(nullptr);
+		} else {
+			// besides the prefix, variable size sorting columns are also fully serialized, along with offsets
+			// we have to assume a large variable size, otherwise a single large variable entry may not fit in a block
+			// 1 << 23 = 8MB
```

Comment:
What happens if a string comes that doesn't fit into an 8MB block?

Reply:
Then we're screwed. 8MB is enough for the entirety of the bible, though.

I can add code to dynamically allocated a larger block in RowChunk::Build to deal with larger strings if we want

Reply:
> I can add code to dynamically allocated a larger block in RowChunk::Build to deal with larger strings if we want

I think this will be required at some point, but it's not necessary for this PR yet, as long as a proper error is thrown when this happens (test case!).

Never underestimate people's ability to generate obscene outliers :) Especially as part of payloads, you can think of e.g. pickled Python objects being pushed in the database which might be gigantic.

Reply:
There is a test in `test/sql/storage/test_big_storage.cpp` where we generate a single huge string using concatenations (`"Test storing big strings"`), for inspiration :)